### PR TITLE
Add DataStore for new revocation charts

### LINKED ISSUE #657: [API-FE] Build a query string based on the current filter state,
**What needs to be done? Why does it need to be done?**
Build a query string based on the current filter state. Add the query string to the API request URI.

Match the params that were established in #653

**Additional context**
Add any other context about the task here.

### LINKED ISSUE #686: [API-FE] Implement that datastore for the API data
**What needs to be done? Why does it need to be done?**
We need to implement a datastore for the API data so that every chart component receives its data as a prop. This is so we can improve the memory, since all anonymous functions defined in the same scope as the `useChartData` hook (which defines apiData) ends up holding onto a reference of the large data object. This solution will separate those definitions and should lower the retained memory size. 

**Additional context**
Add any other context about the task here.

### PR DESCRIPTION
**Pull request by daschi**
## Description of the change

This PR adds two new features:

- Implement a DataStore for the MO and PA dashboard charts so that data fetching and state management is taken out of the components and out of useEffect hooks. 
- Request a new dataset on every filter change

There is a *lot* of duplication in this PR for each DataStore. I decided to spell everything out at this stage because I'm still learning mobx and wanted to get something working to test the IE11 memory performance before DRYing it up. 

This adds the following mobx data stores:

- DataStore - Parent store that initializes the other datastores and makes them available through the useRootStore context hook
- CaseTableStore
- MatrixStore
- RevocationsOverTimeStore
- RevocationsChartsStore - This is one store for the revocations charts since we only want to store one chart's data at a time in the app state

All of the chart datastores should be more or less the same with some small differences in the `filteredData` getter. The `RevocationsChartsStore` has a slightly different `fetchData` function since it fetches data based on the selected chart.

TODO: 
- [x] Use the DataStore for the DistrictFilter
- [x] IE11 performance testing with production build
- [x] Add tests for the new stores

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #686 
> Closes #657 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: src/RootStore/DataStore/RevocationsOverTime.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    reaction(
+      () => getQueryStringFromFilters(this.queryFilters),
+      (queryString) => this.fetchData(queryString)
+    );
+  }
+
+  get queryFilters() {
+    return Object.fromEntries(toJS(this.filtersStore.filters));
+  }
```
Comment (macfarlandian) [inline src/RootStore/DataStore/RevocationsOverTime.js:88]: why doesn't this just call `getQueryStringFromFilters`?

File: src/RootStore/RootStore.js [hunk 2]
```diff
@@ -15,24 +15,45 @@
...
 
+  auth0Context = observable.map({ loading: true });
+
   constructor() {
-    makeObservable(this, { filters: computed, currentTenantId: computed });
+    makeObservable(this, {
+      filters: computed,
+      currentTenantId: computed,
+      setAuth0Context: action,
+    });
```
Comment (macfarlandian) [inline src/RootStore/RootStore.js:41]: i think you should include `getTokenSilently` here too as a `computed`?

File: src/api/metrics/urlHelpers.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// =============================================================================
+import qs from "qs";
+
+/**
+ *
+ * @param {Object} filters - The filter values used to construct the query string to request metric data
+ * @param {string} filters.chargeCategory - A charge category or "All"
+ * @param {Array} filters.district - District IDs or "All"
+ * @param {string} filters.metricPeriodMonths - The number of months in the time period
+ * @param {string} filters.supervisionType - Supervision Type of "All"
```
Comment (macfarlandian) [inline src/api/metrics/urlHelpers.js:25]: *or

File: src/components/charts/new_revocations/CaseTable/CaseTable.js [hunk 4]
```diff
@@ -29,29 +28,23 @@ import {
...
 import { useRootStore } from "../../../../StoreProvider";
 import { METRIC_PERIOD_MONTHS } from "../../../../constants/filterTypes";
+import { dataStorePropTypes } from "../../propTypes";
 
 export const CASES_PER_PAGE = 15;
 
-const CaseTable = ({ dataFilter }) => {
-  const { currentTenantId, filtersStore } = useRootStore();
+const CaseTable = ({ dataStore }) => {
+  const { filtersStore } = useRootStore();
```
Comment (macfarlandian) [inline src/components/charts/new_revocations/CaseTable/CaseTable.js:40]: I don't quite understand why one store comes in as a prop while the other is read from the context? (same question applies to all of the other components that do this)

if nothing else it is making this code hard to follow because it is not obvious which store is being used where

File: src/components/charts/new_revocations/CaseTable/CaseTable.js [hunk 5]
```diff
@@ -29,29 +28,23 @@ import {
...
-  const { isLoading, isError, apiData } = useChartData(
-    `${currentTenantId}/newRevocations`,
-    "revocations_matrix_filtered_caseload"
-  );
-
   const sortedData = useMemo(() => {
-    return dataFilter(apiData || []).sort(comparator);
-  }, [dataFilter, apiData, comparator]);
+    return dataStore.filteredData.sort(comparator);
+  }, [dataStore.filteredData, comparator]);
```
Comment (macfarlandian) [inline src/components/charts/new_revocations/CaseTable/CaseTable.js:47]: I wonder whether it would be cleaner for sorting to be done on the datastore instead of here? mutating the `computed` array in this component feels like it could be bug-prone or at the very least somewhat obscure

File: src/components/charts/new_revocations/CaseTable/CaseTable.js (hunk 4 above)
Comment (daschi) [reply src/components/charts/new_revocations/CaseTable/CaseTable.js:40]: @jovergaag After chatting on slack, we decided to move these into the components.

File: src/components/charts/new_revocations/CaseTable/CaseTable.js (hunk 5 above)
Comment (daschi) [reply src/components/charts/new_revocations/CaseTable/CaseTable.js:47]: For now I updated this to copy the data before calling sort on it. I think this PR is already pretty big so would like to separate out thinking through moving this to the datastore in a different branch.

**Comment by macfarlandian** [comment]
@daschi are the TODOs in the description things you are planning to add to this PR? just wondering whether I should wait for more changes before reviewing again

File: src/RootStore/DataStore/CaseTable.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+  isError = false;
+
+  apiData = [];
+
+  filteredData = [];
+
+  metadata = {};
+
+  auth0Context = observable.map({ loading: true });
```
Comment (jovergaag) [inline src/RootStore/DataStore/CaseTable.js:47]: This is coming from the `rootStore` so I don't think we need it as a a class attribute anymore.

File: src/RootStore/DataStore/Matrix.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      () => this.fetchData(this.queryFilters)
+    );
+
+    reaction(
+      () => this.queryFilters,
+      (queryString) => this.fetchData(queryString)
+    );
+  }
+
+  get queryFilters() {
```
Comment (jovergaag) [inline src/RootStore/DataStore/Matrix.js:76]: This is never different per data store right? Should we put this on the parent `DataStore` class?

File: src/RootStore/DataStore/RevocationsCharts.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    );
+
+    reaction(
+      () => this.selectedChart,
+      () => this.fetchData(this.queryFilters)
+    );
+
+    reaction(
+      () => this.queryFilters,
+      (queryString) => this.fetchData(queryString, false)
```
Comment (jovergaag) [inline src/RootStore/DataStore/RevocationsCharts.js:89]: I think we should set `isLoading` to true here. When I was playing around with this, I changed a filter, and during the time that the fetch was happening in the background the chart didn't change - then all of a sudden it changed. This might be confusing to a user because they may not understand that the chart changed in response to the filter change. I think it would be more clear (on all charts) to set `isLoading` to true when the fetch is initiated.

@sychang Thoughts on this? Should we show the loading indicator in the chart if a filter is changed such that it is requesting new data?

File: src/RootStore/DataStore/CaseTable.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    });
+
+    this.rootStore = rootStore;
+
+    when(
+      () => !get(this.rootStore.auth0Context, "loading"),
+      () => this.fetchData(this.queryFilters)
+    );
+
+    reaction(
```
Comment (jovergaag) [inline src/RootStore/DataStore/CaseTable.js:69]: I think this can be an `autorun` instead of `reaction`, for all data stores.

File: src/RootStore/DataStore/RevocationsOverTime.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    });
+
+    return filterOptimizedDataFormat({
+      apiData: data,
+      metadata,
+      filterFn: dataFilter,
+    });
+  }
+
+  get districts() {
```
Comment (jovergaag) [inline src/RootStore/DataStore/RevocationsOverTime.js:132]: I find this a little confusing to keep this in this store. Ultimately it doesn't matter which metric file we're pulling the list of districts from, but I think it would make more sense to store this list elsewhere.  

We have a feature coming down the pipe where we will receive a file ` supervision_location_ids_to_names.json` which will list all the district numbers and map them to their name to be displayed in the filter. When we implement that we will use that as the list of districts for the filter. Until then, parsing them from this metric file is fine, lets just let it live elsewhere.

What about in the `FiltersStore`?

File: src/RootStore/RootStore.js [hunk 11]
```diff
@@ -15,24 +15,46 @@
...
     this.filtersStore = new FiltersStore({ rootStore: this });
+    this.dataStore = new DataStore({ rootStore: this });
+  }
+
+  setAuth0Context() {
+    const auth0Context = useAuth0();
+    this.auth0Context.merge(auth0Context);
+  }
+
+  get getTokenSilently() {
```
Comment (jovergaag) [inline src/RootStore/RootStore.js:56]: Why can't we just access this directly from `rootStore.auth0Context`?

File: src/api/metrics/urlHelpers.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+// =============================================================================
+import qs from "qs";
+
+/**
+ *
+ * @param {Object} filters - The filter values used to construct the query string to request metric data
+ * @param {string} filters.chargeCategory - A charge category or "All"
+ * @param {Array} filters.district - District IDs or "All"
+ * @param {string} filters.metricPeriodMonths - The number of months in the time period
+ * @param {string} filters.supervisionType - Supervision Type or "All"
```
Comment (jovergaag) [inline src/api/metrics/urlHelpers.js:25]: We're missing `supervisionLevel` in this list. MO and PA have different filters enabled 🎊 .

File: src/components/Revocations.js [hunk 13]
```diff
@@ -108,92 +98,23 @@ const Revocations = () => {
...
-              dataFilter={matchesAllFilters({
-                filters,
-                skippedFilters: [DISTRICT],
-              })}
-              timeDescription={timeDescription}
-            />
-          </ErrorBoundary>
-        }
-      />
+      <RevocationCharts timeDescription={timeDescription} />
```
Comment (jovergaag) [inline src/components/Revocations.js:113]: 😍

Comment (jovergaag) [reply src/components/Revocations.js:113]: I think we lost our `ErrorBoundary` around these charts.

File: src/components/charts/new_revocations/CaseTable/CaseTable.js [hunk 14]
```diff
@@ -29,43 +28,31 @@ import {
...
-
   const [page, setPage] = useState(0);
   const { sortOrder, toggleOrder, comparator } = useSort();
 
-  const { isLoading, isError, apiData, unflattenedValues } = useChartData(
-    `${currentTenantId}/newRevocations`,
-    "revocations_matrix_filtered_caseload",
-    false
-  );
+  const filteredData = store.filteredData.slice();
```
Comment (jovergaag) [inline src/components/charts/new_revocations/CaseTable/CaseTable.js:45]: Why is this necessary?

File: src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js [hunk 15]
```diff
@@ -78,19 +113,14 @@ const RevocationCharts = ({
         ))}
       </div>
       <div className="RevocationCharts__chart">
-        {renderSelectedChartSingularLoad()}
+        <ErrorBoundary>{renderSelectedChartSingularLoad()}</ErrorBoundary>
```
Comment (jovergaag) [inline src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js:116]: Oh there it is. Nevermind!! :)

File: src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js [hunk 16]
```diff
@@ -32,31 +42,56 @@ const CHARTS = [
...
-  officerChart,
-  violationChart,
-  genderChart,
-  raceChart,
-  districtChart,
-}) => {
-  const [selectedChart, setSelectedChart] = useState(CHARTS[0]);
+const RevocationCharts = ({ timeDescription }) => {
+  const { dataStore } = useRootStore();
+  const store = dataStore.revocationsChartsStore;
```
Comment (jovergaag) [inline src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js:47]: Is there a benefit to accessing the store in this parent and passing it down to the chart rather than accessing it directly in the chart?

File: src/components/charts/new_revocations/RevocationsByOfficer/RevocationsByOfficer.js [hunk 17]
```diff
@@ -80,7 +76,7 @@ const RevocationsByOfficer = ({ dataFilter, timeDescription }) => {
           />
         );
       }}
-      generateChartData={createGenerateChartData(dataFilter)}
+      generateChartData={createGenerateChartData(dataStore)}
```
Comment (jovergaag) [inline src/components/charts/new_revocations/RevocationsByOfficer/RevocationsByOfficer.js:79]: Eventually we should generate this chart data in the data store and pass the transformed data to the chart. For another day...

**Review by jovergaag — CHANGES_REQUESTED** [review]
Holy cow this is so nice! Those components are looking so fresh and clean!

I found a few little issues and have a few questions. But this is getting really close I think!

File: src/RootStore/DataStore/DataStore.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+import CaseTableStore from "./CaseTable";
+
+export default class DataStore {
+  rootStore;
+
+  revocationsOverTimeStore;
+
+  matrixStore;
+
+  revocationsCharts;
```
Comment (jovergaag) [inline src/RootStore/DataStore/DataStore.js:31]: I find the plural `revocationsCharts` to be a little misleading, thinking that it would have the data for all the charts. Maybe just `revocationsChart`?

File: src/RootStore/DataStore/RevocationsCharts.js (hunk 8 above)
Comment (sychang) [reply src/RootStore/DataStore/RevocationsCharts.js:89]: Yes, let's add the loading indicator!

**Comment by daschi** [comment]
@macfarlandian I did the IE11 performance testing last Friday, and I'll spend this morning adding the tests for the new data stores and responding to Jen's comments. I'll re-request a review when that's ready! Thanks for checking

File: src/RootStore/DataStore/Matrix.js (hunk 7 above)
Comment (daschi) [reply src/RootStore/DataStore/Matrix.js:76]: Yeah I would like to DRY up the stores, but right now the `DataStore` is not a parent class, it's acting as a "root store" for all of the data stores. I think I could create another store as an abstract class like `BaseDataStore` that will be the parent to the other stores and have all of the shared attributes.

File: src/RootStore/DataStore/CaseTable.js (hunk 9 above)
Comment (daschi) [reply src/RootStore/DataStore/CaseTable.js:69]: Ah okay, that makes sense. I would remove the `when` and the `reaction` and replace it with `autorun` that checks the auth0Context condition.

File: src/RootStore/RootStore.js (hunk 11 above)
Comment (daschi) [reply src/RootStore/RootStore.js:56]: I think we can, I was probably overthinking it :)

File: src/components/charts/new_revocations/CaseTable/CaseTable.js (hunk 14 above)
Comment (daschi) [reply src/components/charts/new_revocations/CaseTable/CaseTable.js:45]: I added this to avoid directly mutating the observable data on line 46 with the sort. @macfarlandian mentioned we should move the sorting to the data store. I started going down that path but it seemed a little more complicated and figured this PR was already getting pretty big that we could tackle that in a separate issue.

File: src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js (hunk 15 above)
Comment (daschi) [reply src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js:116]: Yeah, but you make a good point that it's a little hidden. I'm going to move this back into the `Revocations` component so it matches all the other components' pattern.

File: src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js (hunk 16 above)
Comment (daschi) [reply src/components/charts/new_revocations/RevocationCharts/RevocationCharts.js:47]: I kept it here since we need the `setSelectedChart` and `selectedChart` states in this parent component, but I'm open to also connecting each of the child component charts directly to the store too since mobx says it's better to access the data closer to the rendering component.

File: src/RootStore/DataStore/RevocationsOverTime.js (hunk 10 above)
Comment (daschi) [reply src/RootStore/DataStore/RevocationsOverTime.js]: Oh okay, in that case what do you think about creating a `DistrictsStore`, that will eventually fetch the new json file?

File: src/RootStore/RootStore.js (hunk 11 above)
Comment (jovergaag) [reply src/RootStore/RootStore.js:54]: Oh I see, ok. The `UserStore` removes the useAuth0 context so we will prob have to make adjustments depending on who merges first!

**Comment by daschi** [comment]
This is ready for another review. I pulled out the shared properties and methods into a `BaseDataStore` superclass and added a test for that class. I think there's still some more tests that could be added for each DataStore's `filterData` function, which I could look into tomorrow (or move onto a new issue?). Let me know what you think..

File: src/components/charts/new_revocations/CaseTable/CaseTable.js (hunk 14 above)
Comment (jovergaag) [reply src/components/charts/new_revocations/CaseTable/CaseTable.js:45]: As we've discussed before, I think we should do all the data prep in the store and only send formatted/sorted/filtered data to the components. One day we will be there :)

File: src/components/charts/new_revocations/ToggleBar/DistrictFilter.js [hunk 19]
```diff
@@ -93,9 +77,9 @@ const DistrictFilter = () => {
         isSearchable
       />
     );
-  }, [district, isLoading, apiData, filtersStore, filters]);
+  };
 
-  return <FilterField label="District">{select}</FilterField>;
+  return <FilterField label="District">{getSelectElement()}</FilterField>;
```
Comment (jovergaag) [inline src/components/charts/new_revocations/ToggleBar/DistrictFilter.js:82]: Oh nice. I was wondering what was wrong with the district filter. Thanks for fixing it!!

File: src/RootStore/DataStore/BaseDataStore.js [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+      eagerExpand: true,
+    });
+
+    this.file = file;
+
+    this.rootStore = rootStore;
+
+    autorun(() => {
+      if (!get(this.rootStore.auth0Context, "loading")) {
+        this.fetchData(this.queryFilters);
```
Comment (jovergaag) [inline src/RootStore/DataStore/BaseDataStore.js:71]: One thing that is a little wonky with this (only in IE11 - Chrome is fine) - when I'm on the profile page and switch from one Lantern tenant to the other...the page becomes unresponsive for 20 seconds or so (depending on if switching to PA or MO) while the data is fetched and loaded. This means the user can't navigate back to the dashboard because the "Lantern" button is disabled for a while.

However, only Recidiviz and CSG users are the ones who are able to switch tenants...so maybe this isn't a big deal? I'm inclined to say not a big deal for now...CSG users are not on IE11 (at least to my knowledge). @sychang Does this seem like something we need to resolve? 

We can kick off the fetch when the user navigates back to the dashboard page to make IE11 happy...but it is kind of nice to have that fetching happening in the background for non-IE11 browsers so that when the user gets to the dashboard the data/charts are ready to go.

**Review by jovergaag — CHANGES_REQUESTED** [review]
THIS IS USEABLE IN IE11!!! I actually used the dashboard in IE11- filtered and switched charts for at least 10 mins and it never crashed. Well done!!!

Just one issue with `eagerExpand`, but its oh so close!

File: src/RootStore/DataStore/BaseDataStore.js (hunk 20 above)
Comment (sychang) [reply src/RootStore/DataStore/BaseDataStore.js:71]: This is absolutely not a big deal, since no one on Recidiviz nor CSG are using IE11

Comment (jessex) [reply src/RootStore/DataStore/BaseDataStore.js:71]: Just flying by with a quick comment: sounds like something to defer to some later point, so let's just make sure we file a task so we don't forget!

File: src/RootStore/DataStore/RevocationsOverTimeStore.js [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    super({ rootStore, file: `revocations_matrix_by_month` });
+    makeObservable(this, {
+      fetchData: flow,
+      expandedData: observable.shallow,
+      // TODO: Remove once we get a separate districts file
+      districts: computed,
+    });
+  }
+
+  *fetchData(queryString = "") {
```
[the remark this answers is not in the document]
Comment (daschi) [reply src/RootStore/DataStore/RevocationsOverTimeStore.js:45]: This needs to also store the expanded data to create the `districts` data. I'm going to open a separate issue to create a `DistrictsStore` as a fast follow to this.

File: src/RootStore/__tests__/BaseDataStore.test.js [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+        expect(baseStore.isError).toEqual(false);
+      });
+
+      it("sets apiData", () => {
+        expect(baseStore.apiData).toEqual([["0"], ["0"]]);
+      });
+    });
+  });
+
+  describe("when API responds with an error", () => {
```
[the remark this answers is not in the document]
Comment (daschi) [reply src/RootStore/__tests__/BaseDataStore.test.js:103]: I added a note to that issue that data fetching is coming from the data stores now for new revocations charts.

File: src/api/metrics/__tests__/urlHelpers.test.js [hunk 23]
```diff
@@ -0,0 +1,43 @@
+// Recidiviz - a data platform for criminal justice reform
```
[the remark this answers is not in the document]
Comment (daschi) [reply src/api/metrics/__tests__/urlHelpers.test.js:1]: Do you mean `getQueryStringFromFilters`? Yeah I can move it into the DataStore helpers.

File: src/components/charts/new_revocations/CaseTable/CaseTable.js (hunk 14 above)
Comment (daschi) [reply src/components/charts/new_revocations/CaseTable/CaseTable.js:45]: I can open a issue for this

File: src/RootStore/DataStore/BaseDataStore.js (hunk 20 above)
Comment (daschi) [reply src/RootStore/DataStore/BaseDataStore.js:71]: I can open an issue for this

File: src/api/metrics/__tests__/urlHelpers.test.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (jovergaag) [reply src/api/metrics/__tests__/urlHelpers.test.js]: Yes thats what I meant. Thanks for being psychic :)

**Review by jovergaag — APPROVED** [review]
You're a hero 🌈

**Review by macfarlandian — APPROVED** [review]
🎉
